File: packages/wasmvm/wasmlib/go/wasmlib/assets.go
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package wasmlib

import (
	"sort"

	"github.com/iotaledger/wasp/packages/wasmvm/wasmlib/go/wasmlib/wasmtypes"
)

type TokenAmounts map[wasmtypes.ScTokenID]wasmtypes.ScBigInt

type ScAssets struct {
	BaseTokens   uint64
	NativeTokens TokenAmounts
	NftIDs       map[wasmtypes.ScNftID]bool
}

func NewScAssets(buf []byte) *ScAssets {
	assets := &ScAssets{}
	if len(buf) == 0 {
		return assets
	}

	dec := wasmtypes.NewWasmDecoder(buf)
	empty := wasmtypes.BoolDecode(dec)
	if empty {
		return assets
	}

	assets.BaseTokens = wasmtypes.Uint64Decode(dec)

	size := wasmtypes.Uint16Decode(dec)
	if size > 0 {
		assets.NativeTokens = make(TokenAmounts, size)
		for ; size > 0; size-- {
			tokenID := wasmtypes.TokenIDDecode(dec)
			assets.NativeTokens[tokenID] = wasmtypes.BigIntDecode(dec)
		}
	}

	size = wasmtypes.Uint16Decode(dec)
	if size > 0 {
		assets.NftIDs = make(map[wasmtypes.ScNftID]bool)
	}
	for ; size > 0; size-- {
		nftID := wasmtypes.NftIDDecode(dec)
		assets.NftIDs[nftID] = true
	}
	return assets
}

func (a *ScAssets) Balances() ScBalances {
	return ScBalances{assets: a}
}

func (a *ScAssets) Bytes() []byte {
	if a == nil {
		return []byte{}
	}

	enc := wasmtypes.NewWasmEncoder()
	empty := a.IsEmpty()
	wasmtypes.BoolEncode(enc, empty)
	if empty {
		return enc.Buf()
	}

	wasmtypes.Uint64Encode(enc, a.BaseTokens)

	wasmtypes.Uint16Encode(enc, uint16(len(a.NativeTokens)))
	for _, tokenID := range a.TokenIDs() {
		wasmtypes.TokenIDEncode(enc, *tokenID)
		wasmtypes.BigIntEncode(enc, a.NativeTokens[*tokenID])
	}

	wasmtypes.Uint16Encode(enc, uint16(len(a.NftIDs)))
	for nftID := range a.NftIDs {
		wasmtypes.NftIDEncode(enc, nftID)
	}
	return enc.Buf()
}

func (a *ScAssets) IsEmpty() bool {
	if a.BaseTokens != 0 {
		return false
	}
	for _, val := range a.NativeTokens {
		if !val.IsZero() {
			return false
		}
	}
	return len(a.NftIDs) == 0
}

func (a *ScAssets) TokenIDs() []*wasmtypes.ScTokenID {
	tokenIDs := make([]*wasmtypes.ScTokenID, 0, len(a.NativeTokens))
	for key := range a.NativeTokens {
		// need a local copy to avoid referencing the single key var multiple times
		tokenID := key
		tokenIDs = append(tokenIDs, &tokenID)
	}
	sort.Slice(tokenIDs, func(i, j int) bool {
		return string(tokenIDs[i].Bytes()) < string(tokenIDs[j].Bytes())
	})
	return tokenIDs
}

type ScBalances struct {
	assets *ScAssets
}

func (b *ScBalances) Balance(tokenID *wasmtypes.ScTokenID) wasmtypes.ScBigInt {
	if len(b.assets.NativeTokens) == 0 {
		return wasmtypes.NewScBigInt()
	}
	return b.assets.NativeTokens[*tokenID]
}

func (b *ScBalances) Bytes() []byte {
	if b == nil {
		return []byte{}
	}
	return b.assets.Bytes()
}

func (b *ScBalances) BaseTokens() uint64 {
	return b.assets.BaseTokens
}

func (b *ScBalances) IsEmpty() bool {
	return b.assets.IsEmpty()
}

func (b *ScBalances) NftIDs() map[wasmtypes.ScNftID]bool {
	return b.assets.NftIDs
}

func (b *ScBalances) TokenIDs() []*wasmtypes.ScTokenID {
	return b.assets.TokenIDs()
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

type ScTransfer struct {
	ScBalances
}

// create a new transfer object ready to add token transfers
func NewScTransfer() *ScTransfer {
	return &ScTransfer{ScBalances{assets: NewScAssets(nil)}}
}

// create a new transfer object from a balances object
func ScTransferFromBalances(balances *ScBalances) *ScTransfer {
	transfer := ScTransferFromBaseTokens(balances.BaseTokens())
	for _, tokenID := range balances.TokenIDs() {
		transfer.Set(tokenID, balances.Balance(tokenID))
	}
	for nftID := range balances.NftIDs() {
		transfer.AddNFT(&nftID)
	}
	return transfer
}

// create a new transfer object and initialize it with the specified amount of base tokens
func ScTransferFromBaseTokens(amount uint64) *ScTransfer {
	transfer := NewScTransfer()
	transfer.assets.BaseTokens = amount
	return transfer
}

// create a new transfer object and initialize it with the specified NFT
func ScTransferFromNFT(nftID *wasmtypes.ScNftID) *ScTransfer {
	transfer := NewScTransfer()
	transfer.AddNFT(nftID)
	return transfer
}

// create a new transfer object and initialize it with the specified token transfer
func ScTransferFromTokens(tokenID *wasmtypes.ScTokenID, amount wasmtypes.ScBigInt) *ScTransfer {
	transfer := NewScTransfer()
	transfer.Set(tokenID, amount)
	return transfer
}

func (t *ScTransfer) AddNFT(nftID *wasmtypes.ScNftID) {
	if t.assets.NftIDs == nil {
		t.assets.NftIDs = make(map[wasmtypes.ScNftID]bool)
	}
	t.assets.NftIDs[*nftID] = true
}

func (t *ScTransfer) Bytes() []byte {
	if t == nil {
		return []byte{}
	}
	return t.assets.Bytes()
}

// set the specified tokenID amount in the transfers object
// note that this will overwrite any previous amount for the specified tokenID
func (t *ScTransfer) Set(tokenID *wasmtypes.ScTokenID, amount wasmtypes.ScBigInt) {
	if t.assets.NativeTokens == nil {
		t.assets.NativeTokens = make(TokenAmounts)
	}
	t.assets.NativeTokens[*tokenID] = amount
}

File: packages/wasmvm/wasmlib/as/wasmlib/assets.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {ScDict} from './dict';
import {ScTokenID, tokenIDDecode, tokenIDEncode, tokenIDFromBytes} from './wasmtypes/sctokenid';
import {uint64Decode, uint64Encode} from './wasmtypes/scuint64';
import {bigIntDecode, bigIntEncode, ScBigInt} from './wasmtypes/scbigint';
import {nftIDDecode, nftIDEncode, ScNftID} from './wasmtypes/scnftid';
import {WasmDecoder, WasmEncoder} from './wasmtypes/codec';
import {uint16Decode, uint16Encode} from './wasmtypes/scuint16';
import {boolDecode, boolEncode} from './wasmtypes/scbool';

export class ScAssets {
    baseTokens: u64 = 0;
    nativeTokens: Map<string, ScBigInt> = new Map();
    nftIDs: Set<ScNftID> = new Set();

    public constructor(buf: Uint8Array | null) {
        if (buf === null || buf.length == 0) {
            return this;
        }

        const dec = new WasmDecoder(buf);
        const empty = boolDecode(dec);
        if (empty) {
            return this;
        }

        this.baseTokens = uint64Decode(dec);

        let size = dec.vluDecode(32);
        for (let i: u64 = 0; i < size; i++) {
            const tokenID = tokenIDDecode(dec);
            const amount = bigIntDecode(dec);
            this.nativeTokens.set(ScDict.toKey(tokenID.id), amount);
        }

        size = dec.vluDecode(32);
        for (let i: u64 = 0; i < size; i++) {
            const nftID = nftIDDecode(dec);
            this.nftIDs.add(nftID);
        }
    }

    public balances(): ScBalances {
        return new ScBalances(this);
    }

    public isEmpty(): bool {
        if (this.baseTokens != 0) {
            return false;
        }
        const values = this.nativeTokens.values();
        for (let i = 0; i < values.length; i++) {
            if (!values[i].isZero()) {
                return false;
            }
        }
        return this.nftIDs.size == 0;
    }

    public toBytes(): Uint8Array {
        const enc = new WasmEncoder();
        const empty = this.isEmpty();
        boolEncode(enc, empty);
        if (empty) {
            return enc.buf();
        }

        uint64Encode(enc, this.baseTokens);

        const tokenIDs = this.tokenIDs();
        enc.vluEncode(tokenIDs.length as u64);
        for (let i = 0; i < tokenIDs.length; i++) {
            const tokenID = tokenIDs[i];
            tokenIDEncode(enc, tokenID);
            const mapKey = ScDict.toKey(tokenID.id);
            const amount = this.nativeTokens.get(mapKey);
            bigIntEncode(enc, amount);
        }

        enc.vluEncode(this.nftIDs.size as u64);
        const arr = this.nftIDs.values();
        for (let i = 0; i < arr.length; i++) {
            const nftID = arr[i];
            nftIDEncode(enc, nftID);
        }
        return enc.buf();
    }

    public tokenIDs(): ScTokenID[] {
        const tokenIDs: ScTokenID[] = [];
        const keys = this.nativeTokens.keys().sort();
        for (let i = 0; i < keys.length; i++) {
            const keyBytes = ScDict.fromKey(keys[i]);
            const tokenID = tokenIDFromBytes(keyBytes);
            tokenIDs.push(tokenID);
        }
        return tokenIDs;
    }
}

export class ScBalances {
    assets: ScAssets;

    constructor(assets: ScAssets) {
        this.assets = assets;
    }

    public balance(tokenID: ScTokenID): ScBigInt {
        const mapKey = ScDict.toKey(tokenID.id);
        if (!this.assets.nativeTokens.has(mapKey)) {
            return new ScBigInt();
        }
        return this.assets.nativeTokens.get(mapKey);
    }

    public baseTokens(): u64 {
        return this.assets.baseTokens;
    }

    public isEmpty(): bool {
        return this.assets.isEmpty();
    }

    public nftIDs(): Set<ScNftID> {
        return this.assets.nftIDs;
    }

    public toBytes(): Uint8Array {
        return this.assets.toBytes();
    }

    public tokenIDs(): ScTokenID[] {
        return this.assets.tokenIDs();
    }
}

export class ScTransfer extends ScBalances {
    public constructor() {
        super(new ScAssets(null));
    }

    public static fromBalances(balances: ScBalances): ScTransfer {
        const transfer = ScTransfer.baseTokens(balances.baseTokens());
        const tokenIDs = balances.tokenIDs();
        for (let i = 0; i < tokenIDs.length; i++) {
            const tokenID = tokenIDs[i];
            transfer.set(tokenID, balances.balance(tokenID));
        }
        const nftIDs = balances.nftIDs().values();
        for (let i = 0; i < nftIDs.length; i++) {
            transfer.addNFT(nftIDs[i]);
        }
        return transfer;
    }

    public static baseTokens(amount: u64): ScTransfer {
        const transfer = new ScTransfer();
        transfer.assets.baseTokens = amount;
        return transfer;
    }

    public static nft(nftID: ScNftID): ScTransfer {
        const transfer = new ScTransfer();
        transfer.addNFT(nftID);
        return transfer;
    }

    public static tokens(tokenID: ScTokenID, amount: ScBigInt): ScTransfer {
        const transfer = new ScTransfer();
        transfer.set(tokenID, amount);
        return transfer;
    }

    public addNFT(nftID: ScNftID): void {
        this.assets.nftIDs.add(nftID);
    }

    public set(tokenID: ScTokenID, amount: ScBigInt): void {
        const mapKey = ScDict.toKey(tokenID.id);
        this.assets.nativeTokens.set(mapKey, amount);
    }
}
